src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -13,15 +13,207 @@
 # limitations under the License.
 """Invocation-side implementation of gRPC Asyncio Python."""
 import asyncio
-from typing import Callable, Optional
+from typing import Callable, Dict, Optional
 
+import grpc
 from grpc import _common
 from grpc._cython import cygrpc
 
 SerializingFunction = Callable[[str], bytes]
 DeserializingFunction = Callable[[bytes], str]
 
 
+class AioRpcError(grpc.RpcError):
+    """An RpcError to be used by the asynchronous API."""
+
+    def __init__(self,
+                 code: grpc.StatusCode,
+                 details: Optional[str] = None,
+                 initial_metadata: Optional[Dict] = None,
+                 trailing_metadata: Optional[Dict] = None):
+        self._code = code
+        self._details = details
+        self._initial_metadata = initial_metadata
+        self._trailing_metadata = trailing_metadata
+        super().__init__(self)
+
+    def code(self) -> grpc.StatusCode:
+        """
+        Returns:
+          The `grpc.StatusCode` status code.
+        """
+        return self._code
+
+    def details(self) -> Optional[str]:
+        """
+        Returns:
+          The description of the error.
+        """
+        return self._details
+
+    def initial_metadata(self) -> Optional[Dict]:
+        """
+        Returns:
+          The inital metadata received.
+        """
+        return self._initial_metadata
+
+    def trailing_metadata(self) -> Optional[Dict]:
+        """
+        Returns:
+          The trailing metadata received.
+        """
+        return self._trailing_metadata
+
+
+class Call:
+    """ Object for managing RPC calls, returned when an instance of
+    `UnaryUnaryMultiCallable` object is called.
+    """
+
+    def __init__(self, aio_call: asyncio.Task,
+                 response_deserializer: DeserializingFunction,
+                 aio_call_cancel_status: cygrpc.AioCallCancelStatus):
+        self._exception = None
+        self._response = None
+        self._code = None
+        self._details = None
+        self._initial_metadata = None
+        self._trailing_metadata = None
+        self._cancelled = False
+        self._aio_call = aio_call
+        self._aio_call_cancel_status = aio_call_cancel_status
+        self._response_deserializer = response_deserializer
+
+    def __del__(self):
+        self.cancel()
+
+    def cancel(self) -> bool:
+        """
+        Cancels the ongoing RPC request.
+
+        Returns:
+          True if the RPC can be canceled, False if was already cancelled or terminated.
+        """
+        if self.cancelled() or self.done():
+            return False
```

Comment:
Personally, I don't like the existing way of splitting the state of RPC into multiple boolean. I would appreciate if this PR can make those state deterministic. E.g. has a state enum including `READY`, `ONGOING`, `CANCELLED`, `DONE`, `ABORT` etc.